kimchi/src/circuits/gate.rs

```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
 /// analogously using `joint_combiner`.
 ///
 /// This function computes that combined value.
-pub fn combine_table_entry<'a, F: Field, I: DoubleEndedIterator<Item = &'a F>>(
+pub fn combine_table_entry<
+    'a,
+    F: 'a + Zero + One + Add + Mul + Clone,
+    I: DoubleEndedIterator<Item = &'a F>,
+>(
     joint_combiner: F,
     v: I,
 ) -> F {
-    v.rev().fold(F::zero(), |acc, x| joint_combiner * acc + x)
+    v.rev()
+        .fold(F::zero(), |acc, x| joint_combiner.clone() * acc + x.clone())
 }
 
-impl<F: Field> SingleLookup<F> {
+impl<F: Copy> SingleLookup<F> {
     /// Evaluate the linear combination specifying the lookup value to a field element.
-    pub fn evaluate<G: Fn(LocalPosition) -> F>(&self, eval: G) -> F {
+    pub fn evaluate<K: Zero + Add + Mul<F, Output = K>, G: Fn(LocalPosition) -> K>(
+        &self,
+        eval: G,
+    ) -> K {
         self.value
             .iter()
-            .fold(F::zero(), |acc, (c, p)| acc + *c * eval(*p))
+            .fold(K::zero(), |acc, (c, p)| acc + eval(*p) * *c)
     }
 }
 
 /// A spec for checking that the given vector belongs to a vector-valued lookup table.
 #[derive(Clone, Serialize, Deserialize)]
-pub struct JointLookup<F> {
-    pub table_id: usize,
-    pub entry: Vec<SingleLookup<F>>,
+pub struct JointLookup<SingleLookup> {
+    pub table_id: i32,
+    pub entry: Vec<SingleLookup>,
 }
 
-impl<F: Field> JointLookup<F> {
+pub type JointLookupSpec<F> = JointLookup<SingleLookup<F>>;
+
+impl<F: Zero + One + Add + Mul + Clone> JointLookup<F> {
     // TODO: Support multiple tables
     /// Evaluate the combined value of a joint-lookup.
-    pub fn evaluate<G: Fn(LocalPosition) -> F>(&self, joint_combiner: F, eval: &G) -> F {
-        let mut res = F::zero();
-        let mut c = F::one();
-        for s in self.entry.iter() {
-            res += c * s.evaluate(eval);
-            c *= joint_combiner;
+    pub fn evaluate(&self, joint_combiner: F) -> F {
+        combine_table_entry(joint_combiner, self.entry.iter())
+    }
+}
+
+impl<F: Copy> JointLookup<SingleLookup<F>> {
+    pub fn reduce<K: Zero + Mul<F, Output = K>, G: Fn(LocalPosition) -> K>(
+        &self,
+        eval: &G,
+    ) -> JointLookup<K> {
+        JointLookup {
+            table_id: self.table_id,
+            entry: self.entry.iter().map(|s| s.evaluate(eval)).collect(),
         }
-        res
+    }
+
+    pub fn evaluate<
+        K: Zero + One + Add + Mul + Mul<F, Output = K> + Clone,
+        G: Fn(LocalPosition) -> K,
+    >(
+        &self,
+        joint_combiner: K,
+        eval: &G,
+    ) -> K {
+        self.reduce(eval).evaluate(joint_combiner)
```

Comment:
honestly I find this function really hard to parse.

* we have a `JointLookup<SingleLookup<F>>` for some `F: Zero + One + Add + Mul + Clone`
* we want a `K` for some `K: Zero + One + Add + Mul + Mul<F, Output = K> + Clone`
* in the mean time, we have some eval function that takes a `LocalPosition` and return a `K`

it's just too generic to really grasp what is really going on IMO. Is there a way to make this clearer?

Reply:
interestingly my IDE is also working against myself, pointing me to the wrong reduce function

Reply:
I'm wondering if we can just have code duplication for the different instantiations of these functions, this would make the code much clearer imo and we wouldn't need to pass a closure around (IIUC)